### PR TITLE
automatic picture resizing. fixes #28

### LINKED ISSUE #28: Automatic picture resizing
```
Is it possible to integrate a automatic picture resizing in Polish code? This 
way people can send their pictures without resizing themselves.
The pictures could be resized to 640x480 and after resizing the original could 
be deleted.

My suggestion is to be able to switch this function on or off in 
settings.inc.php


```

Original issue reported on code.google.com by `henricus...@gmail.com` on 31 Jan 2014 at 8:17

### PR DESCRIPTION
**Pull request by wloczynutka**
Ready to merge.
maximum image weight is set to 2.6 MB. Images are automaticly downsize to 640x480, approx. 25-50Kb.
Any comments?

### CODE REVIEW

File: lib/languages/en [hunk 1]
```diff
@@ -436,7 +436,7 @@ what_do_you_find Welcome to {site_name}
...
+what_do_you_find_intro_ro Geocaching este un joc în aer liber prin care juc�?torii încearc�? s�? g�?seasc�? cu ajutorul unui receptor GPS "comori" ascunse de alţi juc�?tori. Aceste "comori" (<i>geocutii</i>) sunt cutii ermetice de diverse forme şi m�?rimi, care conţin mici juc�?rii sau cadouri şi un carneţel (<i>jurnal</i>) în care fiecare juc�?tor înregistreaz�? descoperirea geocutiei. Coordonatele geografice la care se afl�? geocutia sunt p�?strate într-o baz�? de date public�? precum opencaching.ro împreun�? cu o descriere şi instrucţiuni care s�? ajute la g�?sirea ei. <br /> Pe acest site afli despre geocaching în România. Opencaching.ro funcţioneaz�? din <span class="content-title-noshade">aprilie 2014</span> pentru a sprijini şi promova acest joc în România. <br /><br /> Proiectul Opencaching este o abordare <b>open-source</b> a jocului "geocaching" şi se bazeaz�? pe mai multe noduri în diferite ţ�?ri, care fac schimb de informaţii între ele despre geocutiile si utilizatorii fiec�?ruia. <br /> Opencaching.ro funcţioneaz�? pe platforma open-source <span class="content-title-noshade">opencaching-pl</span>. <br /> Acest site nu este în nici un fel afiliat cu Opencaching.com.
```
Comment (wrygiel) [inline lib/languages/en]: Chyba przez przypadek popsułeś unicode.

**Comment by wrygiel** [comment]
BTW - http://stackoverflow.com/questions/2434458/image-resizing-client-side-with-javascript-before-upload-to-the-server - dzięki tej opcji nie musiałbyś mieć żadnego limitu górnego.

**Comment by wloczynutka** [comment]
Po stronie backendu zawsze powinien być limit - wartswę można łatwo 
oszukać. Natomiast zawsze lepiej by było zmieniać rozmiar po stronie 
użytkownika, ale trzeba by fronendowca. bo mi to zajmie zbyt dużo czasu 
i jeszcze będzie źle zrobione :-)

W dniu 2015-07-30 o 23:21, Wojciech Rygielski pisze:

File: lib/languages/en (hunk 1 above)
Comment (wloczynutka) [inline lib/languages/en]: a spróbuj pobrać gałąż na loaklnej maszynie - u mnie plik jest utf-8 bez 
bom. albo warstwa gitlaba daje ciała, albo cos się stało przy 
commitowaniiu. Stawim na to pierwsze.

W dniu 2015-07-30 o 23:18, Wojciech Rygielski pisze:

Comment (wloczynutka) [inline lib/languages/en]: rzeczywiście, jest ansi. Dziwne.. phpStorm daje ciała?

**Comment by andrixnet** [comment]
I have to raise the following points before such a measure is forced
upon users.
Yes, indeed, it is a serious question of hosting resources, but then may
also be considered a loss of functionality.

Automatically resizing (and format converting) images without giving the
user any choice is removing functionality.

I would agree that new images posted to logs may be treated in this
manner. However I would be more cautious and allow more options to cache
creators.
If the procedure is to be applied to both situations, then I would set
different limits for log images and cache listing images.

- a larger image may be significant by it's actual contents and level of
detail.
(/I'm not saying 15MB sized images should be acceptable just like that,
but.../)

- a format such as GIF may be significant from animation point of view.
  a format such as PNG may be significant from transparency point of view.
Resizing both of these may destroy their actual purpose.
GIF animation may be small as width or height but larger file size.
PNG with transparency, especially if it is PNG8 can be rendered badly or
made unusable by resizing.

- images that have extreme aspect ratios may still yield unexpected
results. (for example very wide images or very tall images).
resizing using "fit to square" policy effectively destroy the image's
purpose, resizing to fit designated width may still yield very large
tall images.

- QUESTION: does the user have the option to abort uploading the image 
and be advised to choose a different image hosting service in each of
the following cases:
a) image file size too big (bytes)
b) image pixel width too large
c) image pixel height too large
I mean option presented to user to proceed with upload and accept resize
instead of uploading and resizing automatically, then the user having
the posibility to delete the image, being unsatisfied with the result.

==================
Analyzing the code (diff generated by github) I would like to point out
the following also:

- image width limit is hardcoded. should be configurable.
- image height limit is hardcoded. should be configurable.
- automatic resize as an action should be configurable (enabled for this
node or not)
- consider automatic resize strategy to be "fit to square" by resizing
largest of width&height.
- JPEG compression factor is hardcoded to 75. Should be configurable.
May yield too many compression artifacts.
- $maximumPictureWeight = 2.7; is hardcoded in newpic.php. should be
configurable.

English language note: when talking about images, size refers both to
width&height and bytes. "weight" as terminology is usually meant as a
factor to a parameter determining that parameter's importance among
others. While after some initial confusion one determines the contextual
meaning of "weight", I would recommend to be replaced with "Size" or
"FileSize" for clarity. Thank you.

- newpic.php has some new hardcoded text (moved here from newpic.tpl.php)
- if I am reading the code right, portrait images are resized to fit
height=640 and landscape images are resized to width=480 !!! (not 640)
and exact square images are not resized
- if I am reading the code right, images are automatically converted to
JPEG
- what does title="Uwaga" stand for in newpic.tpl.php?

Thank you.

**Comment by wloczynutka** [comment]
1. Resize do not remove nor disturb any funcionality. Now, user is limited to post 100Kb small image. If he try post larger filesize, error message is displayed. He must downsize picture at home software, then upload again. This feature just make life simpler - user do not need any more downsize images on separate software, image is downsized automaticly.

2. jpg, gif, png are accepted, and not converted into other format.

3. User is still able to attach ANY format and size of picture in description using tinyMce features.

4. Images bellow limit are not processed.

5. picture aspect ratio is not affected

6. Yes, No I see that maybe I missed width and height. and forgot square images. But please test an confirm if images are resized wrong. (you can switch to this branch on your local machine). Then I (or anyone) can correct it.

7. To much configurable things make system not clear. I'm sure many of us at the moment do not know what most of settings in settings.inc.php are used for. And many of us not sure how many of  they really are used. 
If you really interesed witch make it configurable, feel free to do it. We can agree that I will wait for  changes on this branch one week more, and I merge feature on Monday, 10 August. 

8. Please anyone feel free to translate hardcoded things. (They were found and moved here from separate file).

**Comment by andrixnet** [comment]
On 2015-07-31 4:22 PM, Łza z Włóczynutki wrote:
>
> 1.
>
>     Resize do not remove nor disturb any funcionality. Now, user is
>     limited to post 100Kb small image. If he try post larger filesize,
>     error message is displayed. He must downsize picture at home
>     software, then upload again. This feature just make life simpler -
>     user do not need any more downsize images on separate software,
>     image is downsized automaticly.
>
> 2.
>
>     jpg, gif, png are accepted, and not converted into other format.
>
Then what does this line mean?

function save($filename, $image_type = IMAGETYPE_JPEG, $compression = 75, $permissions = null)

My understanding is that it defines default values for optional parameters.
The function contains several if clauses depending on $image_type.
However I could find only one call to this function, and that call only
had one parameter: $filename.
Since $image_type is not propagated and has default value
IMAGETYPE_JPEG, am I correct then in assuming that format conversion
does occur?

If this assumption is correct, then even images that need not be resized
are still processed by same function in the end, and would be converted.

If my assumption is wrong, then I am definitely missing something... (?)
>
> 1.
>
>     User is still able to attach ANY format and size of picture in
>     description using tinyMce features.
>
Indeed. This was never a point.
>
> 1.
>
>     Images bellow limit are not processed.
>
but still saved with same save() function having only filename as
parameter ?
>
> 1.
>
>     picture aspect ratio is not affected
>
Indeed. (it never should be affected)
>
> 1.
>
>     Yes, No I see that maybe I missed width and height. and forgot
>     square images. But please test an confirm if images are resized
>     wrong. (you can switch to this branch on your local machine). Then
>     I (or anyone) can correct it.
>
I will (try). Still working on familiarizing myself with GIT.
>
> 1.
>
>     To much configurable things make system not clear.
>
I would disagree. IMO hardcoded parameters that have no explicit reason
for having a certain value and only that value are a bad practice.
Being a configuration option with a default value (which most people
perhaps will never change) is still much better.
>
> 1.
>
>     I'm sure many of us at the moment do not know what most of
>     settings in settings.inc.php are used for.
>
I have studied the options quite a lot when I installed the site the
first time and fiddled and personalized quite a number of them and
tested the effects of some of them several times before deciding on the
final value.
From sysadmin perspective, it is important to know them/about them and
most important, to have them.
I found many of them of great usefulness. Others I left as default
values, and a few were quite hard to deal with. (the texts defined for
some pages).

Such settings, well grouped and each having a small comment section,
would be best. I did try a few times to do this, however, not wanting to
disturb some ongoing developments, and other constraints, I put the idea
on hold.

I might suggest squid.conf (from squid cache project) as an excellent
reference. >95% of settings are good defaults. All settings are very
well documented with comments in the sample configuration file.

> 1.
>
>     And many of us not sure how many of they really are used.
>
It is true that there may be some settings that are not used, are
deprecated, or are used in some parts of the code, but in other parts
are not. In time, this will gradually be corrected by contributions,
some of which, I hope, will be my own.
>
> 1.
>
>     If you really interesed witch make it configurable, feel free to
>     do it. We can agree that I will wait for changes on this branch
>     one week more, and I merge feature on Monday, 10 August.
>
I will try to experiment with it, as much as possible.
>
> 1.
>
>     Please anyone feel free to translate hardcoded things. (They were
>     found and moved here from separate file).
>
Translation updates are a long overdue task of mine, indeed.
I may also soon add hungarian as language and translation file.

**Comment by wloczynutka** [comment]
> function save($filename, $image_type = IMAGETYPE_JPEG, $compression = 75,
$permissions = null)
> My understanding is that it defines default values for optional
parameters.

Yes, you are right. My mistake. I will correct it. Thanks for fresh point
of view! I aw it, but I did not see it :-)

regards, Lza

2015-08-03 12:18 GMT+02:00 andrixnet <notifications@github.com>:

>
> On 2015-07-31 4:22 PM, Łza z Włóczynutki wrote:
> >
> > 1.
> >
> > Resize do not remove nor disturb any funcionality. Now, user is
> > limited to post 100Kb small image. If he try post larger filesize,
> > error message is displayed. He must downsize picture at home
> > software, then upload again. This feature just make life simpler -
> > user do not need any more downsize images on separate software,
> > image is downsized automaticly.
> >
> > 2.
> >
> > jpg, gif, png are accepted, and not converted into other format.
> >
> Then what does this line mean?
>
> function save($filename, $image_type = IMAGETYPE_JPEG, $compression = 75,
> $permissions = null)
>
> My understanding is that it defines default values for optional parameters.
> The function contains several if clauses depending on $image_type.
> However I could find only one call to this function, and that call only
> had one parameter: $filename.
> Since $image_type is not propagated and has default value
> IMAGETYPE_JPEG, am I correct then in assuming that format conversion
> does occur?
>
> If this assumption is correct, then even images that need not be resized
> are still processed by same function in the end, and would be converted.
>
> If my assumption is wrong, then I am definitely missing something... (?)
> >
> > 1.
> >
> > User is still able to attach ANY format and size of picture in
> > description using tinyMce features.
> >
> Indeed. This was never a point.
> >
> > 1.
> >
> > Images bellow limit are not processed.
> >
> but still saved with same save() function having only filename as
> parameter ?
> >
> > 1.
> >
> > picture aspect ratio is not affected
> >
> Indeed. (it never should be affected)
> >
> > 1.
> >
> > Yes, No I see that maybe I missed width and height. and forgot
> > square images. But please test an confirm if images are resized
> > wrong. (you can switch to this branch on your local machine). Then
> > I (or anyone) can correct it.
> >
> I will (try). Still working on familiarizing myself with GIT.
> >
> > 1.
> >
> > To much configurable things make system not clear.
> >
> I would disagree. IMO hardcoded parameters that have no explicit reason
> for having a certain value and only that value are a bad practice.
> Being a configuration option with a default value (which most people
> perhaps will never change) is still much better.
> >
> > 1.
> >
> > I'm sure many of us at the moment do not know what most of
> > settings in settings.inc.php are used for.
> >
> I have studied the options quite a lot when I installed the site the
> first time and fiddled and personalized quite a number of them and
> tested the effects of some of them several times before deciding on the
> final value.
> From sysadmin perspective, it is important to know them/about them and
> most important, to have them.
> I found many of them of great usefulness. Others I left as default
> values, and a few were quite hard to deal with. (the texts defined for
> some pages).
>
> Such settings, well grouped and each having a small comment section,
> would be best. I did try a few times to do this, however, not wanting to
> disturb some ongoing developments, and other constraints, I put the idea
> on hold.
>
> I might suggest squid.conf (from squid cache project) as an excellent
> reference. >95% of settings are good defaults. All settings are very
> well documented with comments in the sample configuration file.
>
> > 1.
> >
> > And many of us not sure how many of they really are used.
> >
> It is true that there may be some settings that are not used, are
> deprecated, or are used in some parts of the code, but in other parts
> are not. In time, this will gradually be corrected by contributions,
> some of which, I hope, will be my own.
> >
> > 1.
> >
> > If you really interesed witch make it configurable, feel free to
> > do it. We can agree that I will wait for changes on this branch
> > one week more, and I merge feature on Monday, 10 August.
> >
> I will try to experiment with it, as much as possible.
> >
> > 1.
> >
> > Please anyone feel free to translate hardcoded things. (They were
> > found and moved here from separate file).
> >
> Translation updates are a long overdue task of mine, indeed.
> I may also soon add hungarian as language and translation file.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/opencaching/opencaching-pl/pull/191#issuecomment-126690146
> >.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencaching/opencaching-pl/pull/191#issuecomment-127185207>
> .
>



-- 
__________________________________
*Andrzej 'Łza' Woźniak*

*(+48) 798 835 592*
skype praca: *andrzej.wozniak.traveltech*
skype dom: *wloczynutka*

**Comment by harrieklomp** [comment]
This "automatic picture resizing" has a small error with logs where photo's are uploaded. I have added some pictures to show what is happening. I hope this can be corrected in the code :-)
selectie_088
selectie_089

**Pull request merged**